### PR TITLE
Makefile buff-ification

### PR DESCRIPTION
**Pull request by deansheather**
Gets rid of sequential go builds in favor of concurrent ones.

## New targets
```
# builds all slim
build-slim bin

# builds all fat
build-fat build-full build

# to generate site/out/coder.tar.zst
build/coder-slim_${version}_checksums.sha1
build/coder-slim_${version}.tar
build/coder-slim_${version}.tar site/out/coder.tar.zst

# "redirects" to the versioned targets below
build/coder_${arch}_${os}
build/coder-slim_${arch}_${os}

# builds a go binary
build/coder(-slim)?_${version}_${os}_${arch}(.exe)?

# archives
build/coder_${version}_${os}_${arch}.(tar.gz|zip)

# packages (linux only)
build/coder_${version}_linux_${arch}.(apk|deb|rpm)

# docker image "redirects" to versioned targets below
build/coder_linux_${arch}.tag build/coder_linux.tag
push/build/coder_linux_${arch}.tag push/build/coder_linux.tag

# docker image builds and pushes
build/coder_${version}_linux_${arch}.tag build/coder_${version}_linux.tag
push/build/coder_${version}_linux_${arch}.tag push/build/coder_${version}_linux.tag

# helm packages and redirections
build/coder_helm.tgz
build/coder_helm_${version}.tgz
```

### Time savings

For all 7 slim binaries: 30s vs 88s (~66% time saving)

For an entire release build (minus Docker image pushing): 68s vs 250s (~72% time saving)

## TODO:

- [x] Tidy up some makefile spam, maybe make some functions to reduce the spaminess and heavy use of `addprefix`
- [x] Add documentation to the top of each task
- [x] Switch CI to use make instead of calling scripts directly
- [x] Add docker image targets
- [x] Helm package
- [x] Delete unused sh files
- [x] Compare outputs from old builds to new builds to ensure they are the same

### CODE REVIEW

**Review by sreya — APPROVED** [review]
The changes look good although they ultimately make the Makefile less approachable for others to modify. I'm more excited about faster builds though 🚀

File: Makefile [hunk 1]
```diff
@@ -1,59 +1,296 @@
...
+# Push a Docker image.
+$(CODER_ALL_IMAGES_PUSHED): push/%: %
+	image_tag="$$(cat "$<")"
+	docker push "$$image_tag"
+.PHONY: $(CODER_ALL_IMAGES_PUSHED)
+
+# Shortcut for Helm chart package.
+build/coder_helm.tgz: build/coder_helm_$(VERSION).tgz
+	rm -f "$@"
+	ln -s "$$(basename "$<")" "$@"
```
[the remark this answers is not in the document]
Comment (deansheather) [reply Makefile]: good idea! switched to hardlinks

**Review by kylecarbs — APPROVED** [review]
Just two minor things. Unfortunately, this adds complexity to the Makefile, but it was adding complexity to the bash before, so it seems like an overall positive change to me!

File: Makefile [hunk 2]
```diff
@@ -1,59 +1,316 @@
...
+.PHONY: clean
+
+build-slim bin: $(CODER_SLIM_BINARIES)
+.PHONY: build-slim bin
+
+build-fat build-full build: $(CODER_FAT_BINARIES)
+.PHONY: build-fat build-full build
+
+release: $(CODER_FAT_BINARIES) $(CODER_ALL_ARCHIVES) $(CODER_ALL_PACKAGES) $(CODER_ALL_ARCH_IMAGES) build/coder_helm_$(VERSION).tgz
+.PHONY: release
```
[the remark this answers is not in the document]
Comment (deansheather) [reply Makefile:92]: We can potentially in the future after I make the changes to build darwin on linux, but we can't yet.

**Pull request merged**